File: summary_table.R
source("speed.R")

# function to generate a summary table
create_summary_table <- function(data) {
  data_no_geo <- data |>
    st_drop_geometry() |> 
    group_by(month) |> 
    arrange(time) |>
    rename("Month" = month) |> 
    mutate(timestamp = calculate_timediff_seconds(lead(time), time))
  
  table <- summarise(data_no_geo,
                     "Median Interval (min)"= round(median(timestamp, na.rm = TRUE)/60, digits=2),
                     "Mean Interval (min)"= round(mean(timestamp, na.rm = TRUE)/60, digits=2),
                     "Min Interval (s)" = min(timestamp, na.rm=TRUE),
                     "Max Interval (h)" = round(max(timestamp, na.rm=TRUE)/3600, digits=2),
                     "Count (points)" = n(),
                     "Count < 9  min" = sum(timestamp < 540, na.rm = TRUE),
                     "Proportion (%)" =  round(sum(timestamp < 540, na.rm = TRUE) / n(), digits = 2)*100,
                     .groups = "drop")
  
  return(table)
}

File: summary_table.py
import geopandas as gpd
import pandas as pd
from datetime import datetime

def summary_table(data: gpd.GeoDataFrame, datetime_col: datetime) -> pd.DataFrame:
    """
    Create a summary table of the data to get an overview of the data quality.

    Args:
        data: Input GeoDataFrame.
        datetime_col: Input datetime column of type datetime.

    Returns:
        DataFrame: Pandas DataFrame of the table.

    Raises:
        TypeError: If either argument is the wrong data type.
    """
    if not isinstance(data, gpd.GeoDataFrame):
        raise TypeError("Argument data must be gpd.GeoDataFrame")
    elif not isinstance(datetime_col, datetime):
        raise TypeError("Argument datetime_col must be a datetime object")

    data = data.sort_values(by=datetime_col)

    data_no_geo = pd.DataFrame(data.drop(columns=data.geometry.name))
    data_no_geo = data_no_geo.rename(columns={'month': 'Month'})

    table = (data_no_geo.groupby("Month").agg(
            **{"Median Interval (min)": ("timestamp", lambda x: round(x.median() / 60, 2))},
            **{"Mean Interval (min)":   ("timestamp", lambda x: round(x.mean() / 60, 2))},
            **{"Min Interval (s)":      ("timestamp", "min")},
            **{"Max Interval (h)":      ("timestamp", lambda x: round(x.max() / 3600, 2))},
            **{"Count (points)":        ("timestamp", "count")},
            **{"Count < 9 min":         ("timestamp", lambda x: (x < 540).sum())},
            **{"Proportion (%)":        ("timestamp", lambda x: round((x < 540).sum() / len(x), 2) * 100)},
        )
        .reset_index()
    )

    return table


